tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -175,109 +211,127 @@ tfw_sched_hash_get_sg_conn(TfwMsg *msg, TfwSrvGroup *sg)
 static TfwSrvConn *
 tfw_sched_hash_get_srv_conn(TfwMsg *msg, TfwServer *srv)
 {
-	unsigned long msg_hash;
-	size_t tries;
-	TfwHashSrv *srv_cl = srv->sched_data;
+	TfwHashConnList *cl = srv->sched_data;
 
 	/*
-	 * For @srv without connections srv_cl will be NULL, that normally
+	 * For @srv without connections @cl will be NULL, that normally
 	 * does not happen in real life, but unit tests check that case.
 	*/
-	if (unlikely(!srv_cl))
+	if (unlikely(!cl))
 		return NULL;
 
-	msg_hash = tfw_http_req_key_calc((TfwHttpReq *)msg);
-	/* Try several times even if server has just a few connections. */
-	tries = srv_cl->conn_n + 1;
-	while (--tries) {
-		unsigned long best_weight = 0;
-		TfwSrvConn *best_conn = NULL;
-
-		__find_best_conn(&best_conn, srv_cl, &best_weight, msg_hash);
-		if (unlikely(!best_conn))
-			return NULL;
-		if (likely(!tfw_srv_conn_restricted(best_conn)
-			   && !tfw_srv_conn_queue_full(best_conn)
-			   && tfw_srv_conn_get_if_live(best_conn)))
-			return best_conn;
-	}
-	return NULL;
+	return __find_best_conn(msg, cl);
 }
 
 static void
 tfw_sched_hash_del_grp(TfwSrvGroup *sg)
 {
-	size_t si;
-	TfwHashSrvList *sl = sg->sched_data;
+	TfwServer *srv;
+
+	list_for_each_entry(srv, &sg->srv_list, list)
+		if (srv->sched_data) {
+			kfree(srv->sched_data);
+			srv->sched_data = NULL;
+		}
+
+	if (sg->sched_data) {
+		kfree(sg->sched_data);
+		sg->sched_data = NULL;
+	}
+}
+
+static int
+__add_conn(TfwHashConnList *cl, TfwSrvConn *conn, unsigned long hash)
+{
+	ssize_t idx;
+	TfwHashConn new_hcon = {hash, conn};
+
+	idx = __bsearch(hash, cl);
+	if (idx < 0) {
+		/* @cl is empty. */
```

Comment:
It seems you can just check `cl->conn_n` before calling `__bsearch()` - that will make the code simpler. Probably it has sense to check `cl->conn_n` outside of `__basearch()`.